include/grpcpp/impl/codegen/server_callback_impl.h

```diff
@@ -353,835 +398,355 @@ class ServerBidiReactor : public internal::ServerReactor {
   // customization point.
   virtual void InternalBindStream(
       ServerCallbackReaderWriter<Request, Response>* stream) {
-    stream_ = stream;
+    grpc::internal::ReleasableMutexLock l(&stream_mu_);
+    stream_.store(stream, std::memory_order_release);
+    if (send_initial_metadata_wanted_) {
+      stream->SendInitialMetadata();
+      send_initial_metadata_wanted_ = false;
+    }
+    if (read_wanted_ != nullptr) {
+      stream->Read(read_wanted_);
+      read_wanted_ = nullptr;
+    }
+    if (write_and_finish_wanted_) {
+      // Don't perform actual finish actions while holding lock since it could
+      // trigger OnDone that destroys this object including the still-held lock.
+      write_and_finish_wanted_ = false;
+      const Response* write_wanted = write_wanted_;
+      ::grpc::WriteOptions write_options_wanted =
+          std::move(write_options_wanted_);
+      ::grpc::Status status_wanted = std::move(status_wanted_);
+      l.Unlock();
+      stream->WriteAndFinish(write_wanted, std::move(write_options_wanted),
+                             std::move(status_wanted));
+      return;
+    } else {
+      if (write_wanted_ != nullptr) {
+        stream->Write(write_wanted_, std::move(write_options_wanted_));
+        write_wanted_ = nullptr;
+      }
+      if (finish_wanted_) {
+        finish_wanted_ = false;
+        ::grpc::Status status_wanted = std::move(status_wanted_);
+        l.Unlock();
+        stream->Finish(std::move(status_wanted));
+        return;
+      }
+    }
   }
 
-  ServerCallbackReaderWriter<Request, Response>* stream_;
+  grpc::internal::Mutex stream_mu_;
```

Comment:
let's add notes on which fields require locking and which do not

Reply:
Done